README.md
# react-native-touchable-safe

[![npm version](https://img.shields.io/npm/v/react-native-touchable-safe.svg)](https://www.npmjs.com/package/react-native-touchable-safe)
[![Circle CI Status](https://circleci.com/gh/jamesisaac/react-native-touchable-safe.svg?style=shield)](https://circleci.com/gh/jamesisaac/react-native-touchable-safe)
[![license](https://img.shields.io/github/license/jamesisaac/react-native-touchable-safe.svg)](https://opensource.org/licenses/MIT)
[![npm downloads](https://img.shields.io/npm/dm/react-native-touchable-safe.svg)](https://www.npmjs.com/package/react-native-touchable-safe)

A single easy-to-use `<Touchable>` component, which harnesses the power of all React Native's `Touchable*` components.

* Simple API that bridges the differences between RN's various `Touchable*` types.
* A consistent `View` hierarchy, to avoid tricky layout issues when switching between `Touchable*` types.
* Handling the incompatability of ripple customisation on Android API level < 21.

## Motivation

As it stands, `TouchableOpacity` and `TouchableHighlight` wrap their children in a `View`, whereas `TouchableNativeFeedback` and `TouchableWithoutFeedback`
do not.
This can lead to headaches and platform-specific bugs when trying to create advanced Flexbox layouts with different touchable styles on Android/iOS.
An example of this is available here: https://snack.expo.io/ry6kXjX8W

This library makes the situation consistent and easy to reason about:

* `<Touchable>` **always** introduces another view in the hierarchy, which can have its layout customised with `outerStyle`.
* `<Touchable>` **always** must only have one child, which it applies its effect (e.g. opacity) to natively.

## Installation

```bash
$ npm install --save react-native-touchable-safe

# Or, with Yarn
$ yarn add react-native-touchable-safe
```

## Getting started

This component provides a simple API, where alternating the component used per platform is as simple as:

```js
return (
  <Touchable android="native" ios="opacity">
    <MyButton />
  </Touchable>
)
```

In fact, these are the default behaviours, so simply `<Touchable>` is enough to achieve this effect.
The android/ios props only need to be used when deviating from the defaults.

## Props

If you don't want to use the defaults (`TouchableNativeFeedback` on Android and `TouchableOpacity` on iOS), you can specify another type.
Use `all` to set all platforms to the same effect, or `ios` and `android` to differentiate it per platform.

* **`all?`**: `'opacity' | 'highlight' | 'without'`
* **`ios?`**: `'opacity' | 'highlight' | 'without'` - (default: `'opacity'`)
* **`android?`**: `'native' | 'opacity' | 'highlight' | 'without'` - (default: `'native'`)

Some very common behaviours used by all touchable types:

* **`onPress?`**: `() => void`
* **`outerStyle?`**: `Object | number` - Style to pass to the outer `View`
  component which wraps every type of touchable component. Typically used to
  specify things like `<Touchable outerStyle={{ flex: 1 }}>`.
* **`outerProps?`**: `Object` - Similar to `outerStyle`, but lets you set any
  props (although `style` is the main use case).
* **`disabled?`**: `boolean` - Remove any touch functionality and feedback.

Seeing as setting a custom native ripple requires calling `TouchableNativeFeedback.Ripple`, the following top-level convenience props can be used to quickly customise the ripple:

* **`nativeBorderless?`**: `boolean` - For `android="native"`, should the
  ripple effect be borderless.
* **`nativePressColor?`**: `string` - (default: `'rgba(0, 0, 0, .1)'`) - For `android="native"`, what color should the ripple be.

Any props which you only want passed to one type of touchable component can be controlled with the following props.

* **`nativeProps?`**: `Object` - Any props to pass on to a
  `TouchableNativeFeedback` component.
* **`opacityProps?`**: `Object` - Any props to pass on to a
  `TouchableOpacity` component.
* **`highlightProps?`**: `Object` - Any props to pass on to a
  `TouchableHighlight` component.
* **`withoutProps?`**: `Object` - Any props to pass on to a
  `TouchableWithoutFeedback` component.

And finally, anything else will be passed down to all touchable components.

## Examples

### Defaults

NativeFeedback on Android, Opacity on iOS

```js
import React from 'react'
import Touchable from 'react-native-touchable-safe'
import MyButton from './MyButton'

export default () => (
  <Touchable onPress={() => console.log('Pressed')}>
    <MyButton />
  </Touchable>
)
```

### Mixed

A row of different styled buttons, which all behave consistently

```js
import React from 'react'
import { StyleSheet } from 'react-native'
import Touchable from 'react-native-touchable-safe'
import MyButton from './MyButton'

export default ({ disabled }) => (
  <View style={styles.row}>
    {/* Android: native, iOS: highlight */}
    <Touchable
      ios="highlight"
      onPress={() => {
        console.log('Pressed A')
      }}
      outerStyle={styles.touchWrap}
      nativeBorderless
      nativePressColor="rgba(0, 0, 255, .5)"
    >
      <MyButton title="A" />
    </Touchable>

    {/* Both: opacity (50% opacity) */}
    <Touchable
      all="opacity"
      onPress={() => {
        console.log('Pressed B')
      }}
      outerStyle={styles.touchWrap}
      opacityProps={{ activeOpacity: 0.5 }}
    >
      <MyButton title="B" />
    </Touchable>

    {/* Both: no feedback */}
    <Touchable
      all="without"
      onPress={() => {
        console.log('Pressed C')
      }}
      outerStyle={styles.touchWrap}
    >
      <MyButton title="C" />
    </Touchable>

    {/* Both: defaults, disabled based on prop */}
    <Touchable
      onPress={() => {
        console.log('Pressed D')
      }}
      outerStyle={styles.touchWrap}
      disabled={disabled}
    >
      {/* Visual styling of disabled elements handled manually */}
      <MyButton title="D" greyedOut={disabled} />
    </Touchable>
  </View>
)

const styles = StyleSheet.create({
  row: {
    flexDirection: 'row',
    alignItems: 'top',
    height: 200,
  },
  touchWrap: {
    flex: 1,
    height: 100,
  },
})
```


CHANGELOG.md
## 1.1.5

- Add all `TouchableWithoutFeedback` props to the Flow & TypeScript typedefs.
- Instead of replacing the `Touchable` with a `View` when disabled,
  continue rendering the `Touchable` and pass the `disabled` prop down
  (fixes issue with animations being truncated).

## 1.1.4

- Fix type of `outerStyle` in TypeScript typedef

## 1.1.3

- Add TypeScript type declarations in `index.d.ts`

## 1.1.2

- Re-type outerStyle from `mixed` to `any` (until RN style types API stabilises)

## 1.1.1

- Loosen return type of `onPress`

## 1.1.0

- Add an `all` prop for more concisely/explicitly setting both platforms to the same effect.
- Improve Flow types.

## 1.0.2

- Just resorted to `mixed` for `outerStyle` Flow type, as the last attempt was still insufficient to cover everything which RN allows.

## 1.0.1

- Expanded Flow allowed types for `outerStyle`

## 1.0.0

First public release, with formalized API.


js/__tests__/Touchable-test.js
// @flow

import 'react-native'
import { Text, View } from 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import Touchable from '../Touchable'

const Inner = () =>
  <View>
    <Text>Touchable</Text>
  </View>

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Touchable>
        <Inner />
      </Touchable>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})


js/__flowtests__/Touchable-flowtest.js
// @flow

import React from 'react'
import { Text, View } from 'react-native'
import Touchable from '../Touchable'

const Inner = () => (
  <View>
    <Text>Touchable</Text>
  </View>
)

const Correct = (
  <Touchable android="opacity">
    <Inner />
  </Touchable>
)
const CorrectIOS = (
  <Touchable ios="opacity">
    <Inner />
  </Touchable>
)

const TwoChildren = (
  // $FlowExpectedError
  <Touchable ios="opacity">
    <Inner />
    <Inner />
  </Touchable>
)

const BadType = (
  // $FlowExpectedError
  <Touchable android="foo">
    <Inner />
  </Touchable>
)
const BadTypeIOS = (
  // $FlowExpectedError
  <Touchable ios="native">
    <Inner />
  </Touchable>
)

const ValidExtraProp = (
  <Touchable onLongPress={() => {}}>
    <Inner />
  </Touchable>
)
const InvalidExtraProp = (
  // $FlowExpectedError
  <Touchable onFooBar={() => {}}>
    <Inner />
  </Touchable>
)

js/Touchable.js
// @flow

import * as React from 'react'
import {
  Platform,
  TouchableHighlight,
  TouchableNativeFeedback,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
} from 'react-native'

type Props = $ReadOnly<{
  // All touchable types inherit the TouchableWithoutFeedback props
  ...React.ElementProps<typeof TouchableWithoutFeedback>,

  all?: 'opacity' | 'highlight' | 'without',
  ios?: 'opacity' | 'highlight' | 'without',
  android?: 'native' | 'opacity' | 'highlight' | 'without',
  onPress?: () => mixed,
  outerStyle?: any,
  outerProps?: {},
  disabled?: boolean,
  children: React.Element<any>,

  // Specific options for the ripple TouchableNativeFeedback, available as
  // props as a convenience so the caller doesn't need to call
  // TouchableNativeFeedback.Ripple.
  nativeBorderless?: boolean,
  nativePressColor?: string,

  // Props to pass on to a specific view type
  nativeProps?: {},
  opacityProps?: {},
  highlightProps?: {},
  withoutProps?: {},
}>

export default ({
  ios,
  android,
  all,
  onPress = () => {},
  outerStyle,
  outerProps = {},
  children,

  nativeBorderless = false,
  nativePressColor = 'rgba(0, 0, 0, .1)',

  nativeProps = {},
  opacityProps = {},
  highlightProps = {},
  withoutProps = {},

  ...rest
}: Props) => {
  // "ios" and "android" values take priority over "all"
  let type = (Platform.OS === 'android' ? android : ios) || all

  // If no value was provided, fall back to platform defaults
  if (!type) {
    type = Platform.OS === 'android' ? 'native' : 'opacity'
  }

  // Merge outerProps into outerStyle
  outerProps = {
    ...outerProps,
    ...(outerStyle ? { style: outerStyle } : {}),
  }

  if (type === 'opacity') {
    return (
      <TouchableOpacity
        {...outerProps}
        activeOpacity={0.5}
        {...opacityProps}
        {...rest}
        onPress={onPress}
        children={children}
      />
    )
  } else if (type === 'native') {
    // Only on Android Lollipop and above
    if (parseInt(Platform.Version, 10) >= 21) {
      return (
        <View {...outerProps}>
          <TouchableNativeFeedback
            background={TouchableNativeFeedback.Ripple(
              nativePressColor,
              nativeBorderless
            )}
            {...nativeProps}
            {...rest}
            onPress={onPress}
            children={children}
          />
        </View>
      )
    }
    return (
      <View {...outerProps}>
        <TouchableNativeFeedback
          {...nativeProps}
          {...rest}
          onPress={onPress}
          children={children}
        />
      </View>
    )
  } else if (type === 'highlight') {
    return (
      <TouchableHighlight
        {...outerProps}
        accessibilityTraits="button"
        underlayColor="#3C5EAE"
        {...highlightProps}
        {...rest}
        onPress={onPress}
        children={children}
      />
    )
  } else if (type === 'without') {
    return (
      <View {...outerProps}>
        <TouchableWithoutFeedback
          {...withoutProps}
          {...rest}
          onPress={onPress}
          children={children}
        />
      </View>
    )
  } else {
    ;(type: empty)
    throw new Error(`Unknown Touchable type: ${type}`)
  }
}
